lib/db/mysql.js

```diff
@@ -413,92 +407,148 @@ module.exports = function (log, error) {
+  //          s.uaFormFactor, s.lastAccessTime, {  ci.commandName : sc.commandData }
+  // Where  : d.uid = $1 AND d.id = $2
+  var DEVICE = 'CALL device_1(?, ?)'
+
+  MySql.prototype.device = function (uid, id) {
+    return this.readAllResults(DEVICE, [uid, id])
+      .then(rows => aggregateNameValuePairs(rows, 'id', 'commandName', 'commandData', 'availableCommands'))
+      .then(devices => {
+        if (devices.length === 0) {
+          throw error.notFound()
+        }
+        return devices[0]
+      })
   }
 
-  // Select : devices d, deviceCapabilities dc, unverifiedTokens u
+  // Select : devices d, unverifiedTokens u
   // Fields : d.id, d.name, d.type, d.createdAt, d.callbackURL, d.callbackPublicKey,
-  //          d.callbackAuthKey, d.callbackIsExpired, dc.capabilities
+  //          d.callbackAuthKey, d.callbackIsExpired
   // Where  : u.uid = $1 AND u.tokenVerificationId = $2 AND
   //          u.tokenId = d.sessionTokenId AND u.uid = d.uid
-  var DEVICE_FROM_TOKEN_VERIFICATION_ID = 'CALL deviceFromTokenVerificationId_4(?, ?)'
+  var DEVICE_FROM_TOKEN_VERIFICATION_ID = 'CALL deviceFromTokenVerificationId_5(?, ?)'
 
   MySql.prototype.deviceFromTokenVerificationId = function (uid, tokenVerificationId) {
-    return this.readFirstResult(DEVICE_FROM_TOKEN_VERIFICATION_ID, [uid, tokenVerificationId])
-      .then(result => unwrapCapabilities(result, 'capabilities'))
+    return this.readAllResults(DEVICE_FROM_TOKEN_VERIFICATION_ID, [uid, tokenVerificationId])
+      .then(rows => aggregateNameValuePairs(rows, 'id', 'commandName', 'commandData', 'availableCommands'))
+      .then(devices => {
+        if (devices.length === 0) {
+          throw error.notFound()
+        }
+        return devices[0]
+      })
   }
 
-  // Select : sessionTokens t, accounts a, devices d, deviceCapabilities dc, unverifiedTokens ut
+  // Select : sessionTokens t, accounts a, devices d, unverifiedTokens ut
   // Fields : t.tokenData, t.uid, t.createdAt, t.uaBrowser, t.uaBrowserVersion, t.uaOS,
   //          t.uaOSVersion, t.uaDeviceType, t.uaFormFactor, t.lastAccessTime, t.authAt,
   //          a.emailVerified, a.email, a.emailCode, a.verifierSetAt, a.locale,
   //          a.createdAt AS accountCreatedAt,
   //          d.id AS deviceId, d.name AS deviceName, d.type AS deviceType, d.createdAt
   //          AS deviceCreatedAt, d.callbackURL AS deviceCallbackURL, d.callbackPublicKey
   //          AS deviceCallbackPublicKey, d.callbackAuthKey AS deviceCallbackAuthKey,
-  //          d.callbackIsExpired AS deviceCallbackIsExpired,
-  //          dc.capabilities AS deviceCapabilities
+  //          d.callbackIsExpired AS deviceCallbackIsExpired
   //          ut.tokenVerificationId, ut.mustVerify
   // Where  : t.tokenId = $1 AND t.uid = a.uid AND t.tokenId = d.sessionTokenId AND
   //          t.uid = d.uid AND t.tokenId = u.tokenId
-  var SESSION_DEVICE = 'CALL sessionWithDevice_13(?)'
+  var SESSION_DEVICE = 'CALL sessionWithDevice_14(?)'
```

Comment:
So, I'm slightly concerned about doing this additional aggregation work in `sessionWithDevice`.  We call `sessionWithDevice` **a lot** - it's the primary auth method for a bunch of auth-server calls, almost none of which actually need the `availableCommands` bundle.  Is it worth re-introducing a plain old "session" method that doesn't do this extra work, and keeping the more expensive version for when we actually need the full device record?

IIRC we often need *some* of the device metadata, which is why we moved everything to use `sessionWithDevice`, is that accurate?

Reply:
How many `availableCommands` do we reasonably expect in practice? It's worth measuring of course but in the context of a joined query against our `sessionTokens` and `devices` tables, it looks like the time spent inside `aggregateNameValuePairs` might be a drop in the ocean to me.

Reply:
> IIRC we often need *some* of the device metadata, which is why we moved everything to use `sessionWithDevice`, is that accurate?

Yep, I think so. Actually, it's good you asked this question, I think I just found a bug while looking in to it.

Reply:
Probably not that many in practice, so perhaps we can take a wait-and-see approach here and optimize it more holistically if it turns out to be a growing bottleneck in practice.